proton/cert_pinning.py

```diff
@@ -0,0 +1,171 @@
+import base64
+import hashlib
+from ssl import DER_cert_to_PEM_cert
+
+import requests
+from OpenSSL import crypto
+from requests.adapters import HTTPAdapter
+from urllib3.connectionpool import HTTPSConnectionPool
+from urllib3.poolmanager import PoolManager
+from urllib3.util.timeout import Timeout
+
+from .constants import PUBKEY_HASH_DICT
+
+
+class TLSPinningError(requests.exceptions.SSLError):
+    def __init__(self, strerror):
+        self.strerror = strerror
+        super(TLSPinningError, self).__init__(strerror)
+
+
+class TLSPinningHTTPSConnectionPool(HTTPSConnectionPool):
+    """HTTPSConnectionPool that verifies the certificate for each connection"""
+    def __init__(
+        self,
+        host,
+        hash_dict,
+        port=None,
+        strict=False,
+        timeout=Timeout.DEFAULT_TIMEOUT,
+        maxsize=1,
+        block=False,
+        headers=None,
+        retries=None,
+        _proxy=None,
+        _proxy_headers=None,
+        key_file=None,
+        cert_file=None,
+        cert_reqs=None,
+        key_password=None,
+        ca_certs=None,
+        ssl_version=None,
+        assert_hostname=None,
+        assert_fingerprint=None,
+        ca_cert_dir=None,
+        **conn_kw
+    ):
+        self.hash_dict = hash_dict
+        super(TLSPinningHTTPSConnectionPool, self).__init__(
+            host,
+            port,
+            strict,
+            timeout,
+            maxsize,
+            block,
+            headers,
+            retries,
+            _proxy,
+            _proxy_headers,
+            key_file,
+            cert_file,
+            cert_reqs,
+            key_password,
+            ca_certs,
+            ssl_version,
+            assert_hostname,
+            assert_fingerprint,
+            ca_cert_dir,
+            **conn_kw
+        )
+
+    def _validate_conn(self, conn):
+        r = super(TLSPinningHTTPSConnectionPool, self)._validate_conn(conn)
+
+        sock = conn.sock
+
+        pem_certificate = self.get_certificate(sock)
+
+        if self.is_session_secure(pem_certificate, conn, self.hash_dict):
+            return r
+
+    def is_session_secure(self, cert, conn, hash_dict):
+        """Checks if connection is secure"""
+
+        cert_hash = self.extract_hash(cert)
+
+        if not self.validate_hash(cert_hash, hash_dict):
+            # Also generate a report
+            conn.close()
+            raise TLSPinningError("Insecure connection")
+
+        return True
+
+    def validate_hash(self, cert_hash, hash_dict):
+        """Validates the hash agains a known list of hashes/pins"""
+        try:
+            hash_dict[self.host].index(cert_hash)
+        except (ValueError, KeyError, TypeError):
+            return False
+        else:
+            return True
```

Comment:
Usually functions like `validate` don't return anything, they just raise an exception in case the validations fails.
In case you want to have something like `is_valid` then you can put the return True within the `try` block.

Reply:
I do agree with this one:

> Usually functions like validate don't return anything, they just raise an exception in case the validations fails.
In case you want to have something like is_valid then you can put the return True


Although regarding the: 
> within the try block.


Per Python docs it states that [the use of the else clause is better than adding additional code to the try clause because it avoids accidentally catching an exception that wasn’t raised by the code being protected by the try … except statement.](https://docs.python.org/3/tutorial/errors.html) I know that it's not a bad thing here since it's only a return, but it's better to build healthy habits :D